library/pk.c

```diff
@@ -514,6 +514,58 @@ int mbedtls_pk_get_psa_attributes(const mbedtls_pk_context *pk,
             return MBEDTLS_ERR_PK_FEATURE_UNAVAILABLE;
 #endif /* MBEDTLS_PK_RSA_ALT_SUPPORT */
 
+#if defined(MBEDTLS_USE_PSA_CRYPTO)
+        case MBEDTLS_PK_OPAQUE:
+        {
+            psa_key_attributes_t old_attributes = PSA_KEY_ATTRIBUTES_INIT;
+            psa_status_t status = PSA_ERROR_CORRUPTION_DETECTED;
+            status = psa_get_key_attributes(pk->priv_id, &old_attributes);
+            if (status != PSA_SUCCESS) {
+                return MBEDTLS_ERR_PK_BAD_INPUT_DATA;
+            }
+            psa_key_type_t old_type = psa_get_key_type(&old_attributes);
+            switch (usage) {
+                case PSA_KEY_USAGE_SIGN_MESSAGE:
+                case PSA_KEY_USAGE_SIGN_HASH:
+                case PSA_KEY_USAGE_VERIFY_MESSAGE:
+                case PSA_KEY_USAGE_VERIFY_HASH:
+                    if (!(PSA_KEY_TYPE_IS_ECC_KEY_PAIR(old_type) ||
+                          old_type == PSA_KEY_TYPE_RSA_KEY_PAIR)) {
+                        return MBEDTLS_ERR_PK_TYPE_MISMATCH;
+                    }
+                    break;
+                case PSA_KEY_USAGE_DECRYPT:
+                case PSA_KEY_USAGE_ENCRYPT:
+                    if (old_type != PSA_KEY_TYPE_RSA_KEY_PAIR) {
+                        return MBEDTLS_ERR_PK_TYPE_MISMATCH;
+                    }
+                    break;
+                case PSA_KEY_USAGE_DERIVE:
+                    if (!(PSA_KEY_TYPE_IS_ECC_KEY_PAIR(old_type))) {
+                        return MBEDTLS_ERR_PK_TYPE_MISMATCH;
+                    }
+                    break;
+                    break;
+                default:
+                    return MBEDTLS_ERR_PK_TYPE_MISMATCH;
+            }
+            psa_key_type_t new_type = old_type;
+            /* Opaque keys are always key pairs, so we don't need a check
+             * on the input if the required usage is private. We just need
+             * to adjust the type correctly if the required usage is public. */
+            if (usage == PSA_KEY_USAGE_VERIFY_MESSAGE ||
+                usage == PSA_KEY_USAGE_VERIFY_HASH ||
+                usage == PSA_KEY_USAGE_ENCRYPT) {
+                new_type = PSA_KEY_TYPE_PUBLIC_KEY_OF_KEY_PAIR(new_type);
+            }
+            more_usage = psa_get_key_usage_flags(&old_attributes);
```

Comment:
I've changed to requiring that the key must allow `usage`, without which a subsequent call to `mbedtls_pk_import_into_psa` would likely succeed but result in a key that can't do what the user wants.

I don't think this function should open the can of worms of checking the underlying key's algorithm policy. The logic of what would and wouldn't work is complex, and best handled inside `psa_copy_key` (which is the bulk of `mbedtls_pk_import_into_psa` for an opaque input). You have to know about wildcards and about enrollment algorithms, and doing this correctly would require access to functions that are currently `static inline` in `psa_crypto.c`.

There are still usage errors that will lead to successfully creating a key with no good usage, for example if the input key only has the algorithm ECDSA and has a policy with `SIGN | DERIVE` and `mbedtls_pk_import_into_psa` is called with `SIGN`. That can happen if the original key has needlessly permissive usage flags. I don't see an easy way to avoid those.

Reply:
Ok, makes sense. Thanks for explaining!